src/openms/source/ANALYSIS/ID/IDMapper.cpp

```diff
@@ -896,7 +896,7 @@ namespace OpenMS
           // (optionally) check charge state
           if (!ignore_charge_)
           {
-            if (z_p != feat.getCharge()) continue;
+            if (std::abs(z_p) != std::abs(feat.getCharge())) continue;
```

Comment:
This may fix it for pure negative mode but is a) slow and b) will not work for mixed mode.
Would be better if the charges to be checked are defined in the beginning based on the mode.

Reply:
Probably no bottleneck of course but abs is not a cheap function.

Reply:
Probably the `std::abs(z_p)` can be omitted since they the charges in the mzML are always detected as positive numbers?

Reply:
Yes maybe you can leave that out but only if it is in the mzML standard.
Regarding charges from MAD. Yes probably not worth it but I think we should revisit the whole charge handling for metabolomics workflows. I assume there are still many incompatibilities, especially if mixed mode is used. I have no idea how common that is but I received questions about this before.

Reply:
std::abs will not be a performance bottleneck here, but I agree with Julianus that we should understand better how we handle negative charges.

For example, we have tools that are charge aware/are charge sensitive (like the IDMapper in this case). And tools that are charge agnostic, like the feature finders.

Reply:
MAD seems to be the only tool that handles negative charges like that or are there others I'm missing?
One fix could be to make MAD report absolute charge values instead and annotate the polarity. This would also make the min/max charge parameters for MAD clearer and more intuitive. Right now you have to set e.g. min charge = -3 and max charge -1.
But I guess this would mess up older settings too much. What do you think?

Reply:
> std::abs will not be a performance bottleneck here, but I agree with Julianus that we should understand better how we handle negative charges.

This might be worth addressing during the retreat, since it involves a lot of moving pieces. Looking at the MZML OBO charge state there is the absolute number of charges, and there is another CV term to denote whether they are positive or negative. Unfortunately most of our tools don't check this, and just assume positive.

Reply:
Or we have a sprint item at some point to review how charges are used etc. It basically will boil down to tools that use charge but can be charge sign agnostic and those that need the sign.